Rewrite the crypt module and integrate it with ReadWrite

nidunc opened the pull request.

mhoek2 commented on modules/app/read_write.py:
```diff
@@ -15,15 +16,19 @@ def __init__(self) -> None:
                 if file.is_file():
                     # Do not read the file if count_only is true
-                    contents = None if count_only else file.read_bytes()
+                    contents = "".encode() if count_only else file.read_bytes()
```
Using ```python b''``` would've also work.

nidunc replied:
That is certainly more concise, indeed; I’ve changed it to `b""`.

mhoek2 commented on modules/app/read_write.py:
```diff
@@ -15,15 +16,19 @@ def __init__(self) -> None:
+        filename: str
+        contents: bytes
+
     def getFiles(self, path: Path, count_only: bool = False) -> list:
```
For the purpose of type hinting and checking right in the IDE?

nidunc replied:
Basically, yeah; to tell editors what a list of class `FilesDict` contains, indeed, so it’s also easier to work with elements in such a list (when doing `for item in items`, for example). I think Python itself ignores things like this.

nidunc commented:
`Crypt` is now completely integrated with `ReadWrite`.  
I have also drastically cleaned up the `Crypt` module and removed most things that are not necessary anymore.

The `encrypt_files` function now handles everything that is needed for encrypting files and can thus be called from other parts of the program.  
The passwords file is now basically a `.json` file.

I also made some changes to the `ReadWrite` module (kind of out of the scope of this PR, but ah well), like modifying the functions for checking whether there are text files and retrieving text files to only return (whether there are) unencrypted text files (present). I also added new `hasAnyTextFiles` and `getAllTextFiles` functions, which behave the same as `hasTextFiles` and `getTextFiles` previously did.

(And there are some other changes too, but those are the most important ones.)

nidunc commented on modules/app/read_write.py:
```diff
@@ -11,56 +13,114 @@ def __init__(self) -> None:
+        """Check whether there are any text files, regardless of the type."""
+        files = self.getFiles(self.textDir, True)
+
+        self.numShareableFiles = len(files)  # decide whether to leave it here …
```
Should this line be here in the `hasAnyTextFiles` function or in `hasTextFiles`?

mhoek2 replied:
in hasTextFiles, otherwise it would show the 'share_files.py' even if there are no valid txt files.

nidunc replied:
Alright, I’ll remove the declaration from hasAnyTextFiles then and keep it in hasTextFiles.

nidunc commented:
There is an error if the `passwords.secret` files does not exist, which I am still trying to fix, but I’ll get to that next week; so this PR is ‘on hold’ for the time being.

nidunc commented:
> There is an error if the `passwords.secret` files does not exist, which I am still trying to fix, but I’ll get to that next week; so this PR is ‘on hold’ for the time being.

I fixed it, and made some other changes, so it’s now ready to be reviewed again and merged. (Some conflicts will obviously have to be resolved, as this PR is based on an older codebase.)

mhoek2 reviewed (commented):
I have tested the ```test_encrypt()``` and ```test_decrypt``` functions.
And it seems they work fine. Gj.

I wrote two minor comments on a few things that I noticed at first glance.

nidunc commented on .gitignore:
```diff
@@ -1,6 +1,12 @@
+fake_Nick_file.txt
+Generated_keys.txt
+Keys_test.txt
+Picked_keys.txt
```
They are, indeed; this is just in case someone still locally has those files, so they aren’t accidentally included in the repository.

mhoek2 replied:
That shouldn't be the case, because pull requests should be reviewed
